lib/logstash/filters/geoip.rb

```diff
@@ -152,67 +155,98 @@ def filter(event)
+      set_data(event)
       return
     rescue java.net.UnknownHostException => e
       @logger.error("IP Field contained invalid IP address or hostname", :exception => e, :field => @source, :event => event)
-      event[@target] = {}
+      set_data(event)
       return
     rescue Exception => e
       @logger.error("Unknown error while looking up GeoIP data", :exception => e, :field => @source, :event => event)
-      event[@target] = {}
-      return
+      # Dont' swallow this, bubble up for unknown issue
+      raise e
     end
-
+    
     event[@target] = geo_data_hash
 
     filter_matched(event)
   end # def filter
+  
+  def populate_data(event, response, ip_address)
+    country = response.getCountry()
+    subdivision = response.getMostSpecificSubdivision()
+    city = response.getCity()
+    postal = response.getPostal()
+    location = response.getLocation()
+
+    geo_data_hash = Hash.new()
+    
+    # if location is empty, there is no point populating geo data
+    # and most likely all other fields are empty as well
+    if location.getLatitude().nil? && location.getLongitude().nil?
+      tag_unsuccessful_lookup(event)
+      return geo_data_hash
+    end
+    
+    @fields.each do |field|
+      case field
+      when "city_name"
+        geo_data_hash["city_name"] = city.getName()
+      when "country_name"
+        geo_data_hash["country_name"] = country.getName()
+      when "continent_code"
+        geo_data_hash["continent_code"] = response.getContinent().getCode()
+      when "continent_name"
+        geo_data_hash["continent_name"] = response.getContinent().getName()
+      when "country_code2"
+        geo_data_hash["country_code2"] = country.getIsoCode()
+      when "country_code3"
+        geo_data_hash["country_code3"] = country.getIsoCode()
+      when "ip"
+        geo_data_hash["ip"] = ip_address.getHostAddress()
+      when "postal_code"
+        geo_data_hash["postal_code"] = postal.getCode()
+      when "dma_code"
+        geo_data_hash["dma_code"] = location.getMetroCode()
+      when "region_name"
+        geo_data_hash["region_name"] = subdivision.getName()
+      when "region_code"
+        geo_data_hash["region_code"] = subdivision.getIsoCode()
+      when "timezone"
+        geo_data_hash["timezone"] = location.getTimeZone()
+      when "location"
+        geo_data_hash["location"] = [ location.getLongitude(), location.getLatitude() ]
+      when "latitude"
+        geo_data_hash["latitude"] = location.getLatitude()
+      when "longitude"
+        geo_data_hash["longitude"] = location.getLongitude()
+      else
+        raise Exception.new("[#{field}] is not a supported field option.")
+      end
+    end
+    
+    return geo_data_hash
+    
+  end
+  
+  def tag_unsuccessful_lookup(event)
+    @logger.debug? && @logger.debug("IP #{event[@source]} was not found in the database", :event => event)
+    @tag_on_failure.each{|tag| event.tag(tag)}
+  end
+  
+  def set_data(event, geo_data = {}) 
```

Comment:
is this method necessary? 

maybe a better approach is to have a `geo_data` object that is instantiated with `Hash.new` and then gets populated by the above `populate_data` method.